homeassistant/components/overkiz/climate_entities/atlantic_electrical_heater_with_adjustable_temperature_setpoint.py

```diff
@@ -0,0 +1,138 @@
+
+from ..coordinator import OverkizDataUpdateCoordinator
+from ..entity import OverkizEntity
+
+PRESET_AUTO = "auto"
+PRESET_COMFORT1 = "comfort-1"
+PRESET_COMFORT2 = "comfort-2"
+PRESET_FROST_PROTECTION = "frost_protection"
+PRESET_PROG = "prog"
+
+
+# Map Overkiz presets to Home Assistant presets
+OVERKIZ_TO_PRESET_MODE = {
+    OverkizCommandParam.OFF: PRESET_NONE,
+    OverkizCommandParam.FROSTPROTECTION: PRESET_FROST_PROTECTION,
+    OverkizCommandParam.ECO: PRESET_ECO,
+    OverkizCommandParam.COMFORT: PRESET_COMFORT,
+    OverkizCommandParam.COMFORT_1: PRESET_COMFORT1,
+    OverkizCommandParam.COMFORT_2: PRESET_COMFORT2,
+    OverkizCommandParam.AUTO: PRESET_AUTO,
+    OverkizCommandParam.BOOST: PRESET_BOOST,
+    OverkizCommandParam.INTERNAL: PRESET_PROG,
+}
+
+PRESET_MODE_TO_OVERKIZ = {v: k for k, v in OVERKIZ_TO_PRESET_MODE.items()}
+
+# Map Overkiz HVAC modes to Home Assistant HVAC modes
+OVERKIZ_TO_HVAC_MODE = {
+    OverkizCommandParam.ON: HVACMode.HEAT,
+    OverkizCommandParam.OFF: HVACMode.OFF,
+    OverkizCommandParam.AUTO: HVACMode.AUTO,
+    OverkizCommandParam.BASIC: HVACMode.HEAT,
+    OverkizCommandParam.STANDBY: HVACMode.OFF,
+    OverkizCommandParam.INTERNAL: HVACMode.AUTO,
+}
+
+HVAC_MODE_TO_OVERKIZ = {v: k for k, v in OVERKIZ_TO_HVAC_MODE.items()}
+
+
+class AtlanticElectricalHeaterWithAdjustableTemperatureSetpoint(
+    OverkizEntity, ClimateEntity
+):
+    """Representation of Atlantic Electrical Heater (With Adjustable Temperature Setpoint)."""
+
+    _attr_hvac_modes = [*HVAC_MODE_TO_OVERKIZ]
+    _attr_preset_modes = [*PRESET_MODE_TO_OVERKIZ]
+    _attr_temperature_unit = TEMP_CELSIUS
+    _attr_supported_features = (
+        ClimateEntityFeature.PRESET_MODE | ClimateEntityFeature.TARGET_TEMPERATURE
+    )
+
+    def __init__(
+        self, device_url: str, coordinator: OverkizDataUpdateCoordinator
+    ) -> None:
+        """Init method."""
+        super().__init__(device_url, coordinator)
+        self.temperature_device = self.executor.linked_device(2)
+
+    @property
+    def hvac_mode(self) -> str:
+        """Return hvac operation ie. heat, cool mode."""
+        if OverkizState.CORE_OPERATING_MODE in self.device.states:
+            state = self.executor.select_state(OverkizState.CORE_OPERATING_MODE)
+            return OVERKIZ_TO_HVAC_MODE[OverkizCommandParam(state)]
+        if OverkizState.CORE_ON_OFF in self.device.states:
+            state = self.executor.select_state(OverkizState.CORE_ON_OFF)
+            return OVERKIZ_TO_HVAC_MODE[OverkizCommandParam(state)]
+
+        return HVACMode.OFF
+
+    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
+        """Set new target hvac mode."""
+        if OverkizState.CORE_OPERATING_MODE in self.device.states:
+            await self.executor.async_execute_command(
+                OverkizCommand.SET_OPERATING_MODE, HVAC_MODE_TO_OVERKIZ[hvac_mode]
+            )
+        elif hvac_mode == HVACMode.OFF:
```

Comment:
If I understand the code correctly, If the device have the state `CORE_OPERATING_MODE`, we'll never reach the `elif` or the `else` here.

Maybe the first if should check against `HVAC_MODE_TO_OVERKIZ`?

Reply:
Sorry, I don’t get what you mean.

Reply:
In  async_set_hvac_mode, the first if is :  
`if OverkizState.CORE_OPERATING_MODE in self.device.states:`

I think this will be tru everytime becaause every AEH device will have this state defined.  
So every other element on this function won't be used.

What I think the code should looks like (if I understand correctly) for this function should be:  
```
        if hvac_mode == HVACMode.OFF:
            await self.executor.async_execute_command(OverkizCommand.OFF)
        elif hvac_mode in HVAC_MODE_TO_OVERKIZ:
            await self.executor.async_execute_command(
                OverkizCommand.SET_OPERATING_MODE, HVAC_MODE_TO_OVERKIZ[hvac_mode]
            )
```

Reply:
You are right, the conditions on these branches are not on the same topic. This device has been implemented with https://github.com/iMicknl/ha-tahoma/issues/304 and this specific condition by @tillstaff 
IMO we should just implement the method like this:

```
    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
        """Set new target hvac mode."""
        await self.executor.async_execute_command(
            OverkizCommand.SET_OPERATING_MODE, HVAC_MODE_TO_OVERKIZ[hvac_mode]
        )
```

I will see If I can ask to someone credentials to directly test the device.

Reply:
@Quentame Did you check if you own this device 🙏 ?

Reply:
Yes, I have multiple https://www.thermor.fr/produits/radiateur-electrique/radiateur-a-chaleur-douce/ingenio-3 devices:
- 2000W Horizontal
- 1500W Horizontal
- 1250W Horizontal x2

As Thermor is an Atlantic group brand, it works with the Atlantic Cozytouch hub.

Reply:
Do you think you will be able to check/test the `async_set_hvac_mode` method and my suggestion just above?

Reply:
Yes I can test but I'm not sure if I understand what piece of code I should test.

Can you provide me a branch to test on ?

Reply:
@tillstaff Thx for your explanation. It’s make sense. So I have remove all the conditions, and keep only the call to SET_OPERATING_MODE. 

@Quentame I’m talking about this branch. As you own the device, it will be easier for you to test if the behavior is correct.